### PR TITLE
Add summarizeInt/Float/Long/Double, reduce and reduceInPlace as default methods on RichIterable and create Collectors2 with Collector implementations for Eclipse Collections.

### PR DESCRIPTION
**Pull request by donraab**

### CODE REVIEW

File: eclipse-collections-api/src/main/java/org/eclipse/collections/api/RichIterable.java [hunk 1]
```diff
@@ -1564,6 +1574,88 @@
...
+    default Optional<T> reduce(BinaryOperator<T> accumulator)
+    {
+        if (this.isEmpty())
+        {
+            return Optional.empty();
+        }
+        T result = this.injectInto(null, (result1, each) ->
+                result1 == null ? each : accumulator.apply(result1, each));
+        return Optional.of(result);
+    }
```
Comment (motlin) [inline eclipse-collections-api/src/main/java/org/eclipse/collections/api/RichIterable.java]: The method `isEmpty()` can be O(n) on `LazyIterable`s so I think that `reduce()` should be rewritten to only call `each()` or should be left abstract and implemented in subclasses.

**Comment by motlin** [comment]
Could you add information to [RELEASE_NOTE_DRAFT.md](https://github.com/eclipse/eclipse-collections/blob/master/RELEASE_NOTE_DRAFT.md) about the new features?

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/bimap/AbstractBiMap.java [hunk 2]
```diff
@@ -371,7 +371,7 @@ public void forEachKeyValue(Procedure2<? super K, ? super V> procedure)
     @Override
     public void each(Procedure<? super V> procedure)
     {
-        this.getInverse().forEachKey(procedure);
+        this.getDelegate().forEachValue(procedure);
     }
```
Comment (motlin) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/bimap/AbstractBiMap.java]: I thought this was handled separately - I don't remember where that commit lives. There's a second method that needs this change, `each()` and it's a bug-fix so I think it makes sense to fix in an earlier commit with dedicated tests.

File: jmh-tests/src/main/java/org/eclipse/collections/impl/jmh/CollectorJoiningTest.java [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+import org.openjdk.jmh.annotations.OutputTimeUnit;
+import org.openjdk.jmh.annotations.Scope;
+import org.openjdk.jmh.annotations.Setup;
+import org.openjdk.jmh.annotations.State;
+import org.openjdk.jmh.annotations.TearDown;
+
+@State(Scope.Thread)
+@BenchmarkMode(Mode.Throughput)
+@OutputTimeUnit(TimeUnit.SECONDS)
+public class CollectorJoiningTest extends AbstractJMHTestRunner
```
Comment (motlin) [inline jmh-tests/src/main/java/org/eclipse/collections/impl/jmh/CollectorJoiningTest.java]: This test is fine though it seems unrelated to the commit message. Maybe it belongs in a separate commit?

File: unit-tests-java8/src/test/java/org/eclipse/collections/test/RichIterableTestCase.java [hunk 4]
```diff
@@ -885,12 +891,40 @@ default void RichIterable_iterationOrder()
...
+        });
+        assertEquals(expectedIterationOrder, summarizeLongOrder);
+
+        MutableCollection<Integer> summarizeDoubleOrder = this.newMutableForFilter();
+        this.getInstanceUnderTest().summarizeDouble(each -> {
+            summarizeDoubleOrder.add(each);
+            return 0;
+        });
+        assertEquals(expectedIterationOrder, summarizeDoubleOrder);
     }
```
Comment (motlin) [inline unit-tests-java8/src/test/java/org/eclipse/collections/test/RichIterableTestCase.java]: The other methods (reduceInPlace) should be covered here too.

Comment (donraab) [inline unit-tests-java8/src/test/java/org/eclipse/collections/test/RichIterableTestCase.java]: Yes, the other methods are covered in this test case as well.

File: jmh-tests/src/main/java/org/eclipse/collections/impl/jmh/CollectorJoiningTest.java (hunk 3 above)
Comment (donraab) [inline jmh-tests/src/main/java/org/eclipse/collections/impl/jmh/CollectorJoiningTest.java]: I added this test to benchmark any performance difference between reduceInPlace(Collector) and stream().collect(Collector) since the implementation is significantly simpler in the serial reduceInPlace case.  I may expand this benchmark if we implement reduceInPlace on ParallelIterable.

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/bimap/AbstractBiMap.java (hunk 2 above)
Comment (donraab) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/bimap/AbstractBiMap.java]: Yes, I believe you have the code and tests implemented for this.

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/collector/Collectors2.java [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+                StringJoiner::toString,
+                EMPTY_CHARACTERISTICS);
+    }
+
+    public static <T> Collector<T, ?, MutableStack<T>> toStack()
+    {
+        return Collector.of(
+                Stacks.mutable::empty,
+                MutableStack::push,
+                (r1, r2) -> {r2.each(r1::push); return r1;},
```
Comment (motlin) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/collector/Collectors2.java]: Are the combiner portions of each Collector covered by tests? Most of them are obviously correct, but Stack is tricky so I want to be sure it's covered.

File: unit-tests-java8/src/test/java/org/eclipse/collections/test/RichIterableTestCase.java [hunk 6]
```diff
@@ -1590,13 +1591,25 @@ default void RichIterable_makeString_appendString()
...
+
+        assertEquals(
                 "[4/4/4/4/3/3/3/2/2/1]",
                 iterable.makeString("[", "/", "]"));
 
+        assertEquals(
+                iterable.makeString("[", "/", "]"),
+                iterable.reduceInPlace(makeString("[", "/", "]")));
+
         StringBuilder builder1 = new StringBuilder();
```
Comment (motlin) [inline unit-tests-java8/src/test/java/org/eclipse/collections/test/RichIterableTestCase.java]: RichIterable_makeString_appendString is overridden in subclasses of RichIterableTestCase. I'd prefer if this were in a new test method, but it could stay in the same test as long as you also modify the overrides.

Comment (motlin) [inline unit-tests-java8/src/test/java/org/eclipse/collections/test/RichIterableTestCase.java]: RichIterable_makeString_appendString is overridden in subclasses of RichIterableTestCase. I'd prefer if this were in a new test method, but it could stay in the same test as long as you also modify the overrides.

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/collector/Collectors2.java [hunk 7]
```diff
@@ -50,12 +62,40 @@ private Collectors2()
...
+    }
+
+    public static <T> Collector<T, MutableList<T>, MutableStack<T>> toStack()
+    {
+        return Collector.of(
+                Lists.mutable::empty,
+                MutableList::add,
+                MutableList::withAll,
+                Stacks.mutable::ofAll,
                 EMPTY_CHARACTERISTICS);
```
Comment (motlin) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/collector/Collectors2.java]: I see you edited toStack() further and my previous comment doesn't apply anymore, but I'm still curious about test coverage.

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/collector/Collectors2.java (hunk 5 above)
Comment (donraab) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/collector/Collectors2.java]: No, great catch.  I need to add some tests using parallelStream().collect() for each Collector.
